[FEAT] Add state machine for parsing request body

leske42 opened the pull request:
I have turned the function into state machine as requested.

Okay it is not actually a state machine, because it handles everything else too. I figured since we inspect every single character now, we don't need the codec anymore. I can just separate the characters i need into chunk size and chunk body itself while parsing, just like at parsing the request line. But i added also the validation checks which go beyond state machine functionality (like payload too large or not mismatched size), i do not think it would make much sense to separate.

Can be checked with a huge file if you modify the client_max_body_size in the configfile (otherwise you will get `413`). 

`./chunked_upload http://127.0.0.1:8080/upload OTHER_BIG_FILE_2 100  0.42s user 0.37s system 0% cpu 2:11.01 total` - uploading 100mb file took about 2 minutes, idk what to think about this. Don't do it through VSCode though cause it builds some cache of the uploaded file and then the process uses too much memory and gets killed.

leske42 commented:
This closes #116 and #122 
But idk what i think about inspecting every single character, as chunked is usually used when the body is super large so it feels weird to do that. But it's the only way to avoid the split CRLF issue if we dont want some specific hardcoded fix for that. I just feel like this upload is slow.

leske42 commented:
I wanted to change the structure of state machine but now i actually think it's not necessary. I did some performance testing, if i disable the logger the upload is actually quite fast. Immediate with chunks of 1000, and 5 seconds with chunks of 100. Real life chunk size will probably be bigger than that. So i think we are fine for now.

LeaYeh reviewed (changes requested):
We need to change 2 things

leske42 commented:
> We need to change 2 things

I have changed 2 things.

1) Changed `std::string` to `std::vector<char>` for chunkbuf. Wanted `unsigned char` at first but it turns out you cannot `write()` unsigned char so char it is. This does not seem to make a difference however, seems like `std::string` builtins don't actually stop at `\0` byte. Anyway I didn't change it back, better be safe than sorry.

2) I changed the chunk length check to the following method: after we decode the length and pass the first CRLF, i keep processing characters until said length is reached (if it says 1000, i will append 1000, no questions asked). After that, i check for CRLF however and if they cannot provide it, then it's exception time.

Number 2) actually seems to solve the issue why we could not upload video or PDF. They indeed had literal CRLF sequences in the raw data but like what can we do. I have tested now and uploading non-plaintext type should also work well.

LeaYeh commented:
Resolved #122

leske42 commented:
I have put the new things we found into #124
